inc/dashboard/namespace.php

```diff
@@ -144,3 +146,82 @@ function get_plugin_version() : string {
 	}
 	return '';
 }
+
+/**
+ * Intercept block preview requests for block thumbnail service requests.
+ *
+ * @param \WP_Query $query WP Query object.
+ *
+ * @return void
+ */
+function block_preview_check( \WP_Query $query ) : void {
+	$block_id = filter_input( INPUT_GET, 'preview-block-id', FILTER_SANITIZE_NUMBER_INT );
+	$hmac = filter_input( INPUT_GET, 'key' );
+
+	if (
+		! $query->is_main_query()
+		|| empty( $block_id )
+		|| empty( $hmac )
+		|| $hmac !== get_block_thumbnail_request_hmac( $block_id )
+	) {
+		return;
+	}
+
+	$allow_block_thumbnails = is_block_thumbnails_allowed( $block_id );
+
+	if ( ! $allow_block_thumbnails ) {
+		return;
+	}
+
+	$query->set( 'p', $block_id );
+	$query->set( 'post_type', 'wp_block' );
+	$query->set( 'post_status', 'any' );
+
+	add_action( 'template_redirect', __NAMESPACE__ . '\\block_thumbnail_template_override' );
+}
+
+/**
+ * Return whether block thumbnail functionality is allowed.
+ *
+ * @param int $block_id Block ID.
+ *
+ * @return boolean
+ */
+function is_block_thumbnails_allowed( int $block_id = null ) : bool {
```

Comment:
Grammatically the non-plural version reads better as we're just checking a single block.

Reply:
```suggestion
function is_block_thumbnail_allowed( int $block_id = null ) : bool {
```